lib/Transforms/ConvertIfToSelect/ConvertIfToSelect.cpp

```diff
@@ -0,0 +1,72 @@
+#include "lib/Transforms/ConvertIfToSelect/ConvertIfToSelect.h"
+
+#include "mlir/include/mlir/Dialect/SCF/IR/SCF.h"  // from @llvm-project
+#include "mlir/include/mlir/IR/Operation.h"        // from @llvm-project
+#include "mlir/include/mlir/IR/PatternMatch.h"     // from @llvm-project
+#include "mlir/include/mlir/Transforms/GreedyPatternRewriteDriver.h"  // from @llvm-project
+#include "mlir/include/mlir/Transforms/Passes.h"  // from @llvm-project
+
+using namespace mlir;
+using namespace mlir::scf;
+
+namespace mlir {
+namespace heir {
+
+#define GEN_PASS_DEF_CONVERTIFTOSELECT
+#include "lib/Transforms/ConvertIfToSelect/ConvertIfToSelect.h.inc"
+
+struct IfToSelectConversion : OpRewritePattern<IfOp> {
+  using OpRewritePattern<IfOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(IfOp ifOp,
+                                PatternRewriter &rewriter) const override {
+    // Hoist instructions in the 'then' and 'else' regions
+    auto thenOps = ifOp.getThenRegion().getOps();
+    auto elseOps = ifOp.getElseRegion().getOps();
+
+    rewriter.setInsertionPointToStart(ifOp.thenBlock());
+    for (auto &operation : llvm::make_early_inc_range(
+             llvm::concat<Operation>(thenOps, elseOps))) {
+      if (!llvm::isa<scf::YieldOp>(operation)) {
```

Comment:
While we cannot check the necessity of the conversion without the secretness dataflow analysis, I just realized we can probably already thow a hard error here if the operation to be hoisted does not have the `NoMemoryEffect` and `AlwaysSpeculatable` traits (i.e. is `Pure`). 

It'd probably also be a good idea to add a TODO here to extend support to `ConditionallySpeculatable`, which includes operations such as `arith.divui` which is only safely hoistable if we can statically exlcude a zero division.
To do this "properly", first create an issue for the new task, then add `//TODO (#123): ....` (but replacing 123 with your issue's number). This way, the CI system will automatically pick this up and link it with the issue.

Reply:
For reporting an error, use `ifOp.emitError() << "message"; return failure();`

But I think it is also fine to omit the safety checks in the interim. Nothing is going to production any time soon :)

Reply:
Turns out, `isPure(...)` in C++ doesn't actually check for the equivalent components of the `Pure` trait (a tablegen-only thing ) but instead does a proper check which also handles conditionally speculatable ops, so we get this for free :)